### PR TITLE
Support expr when use `typescript_custom_section` attribute

### PR DESCRIPTION
**Pull request by swnb**
for example

```rust
#[wasm_bindgen(typescript_custom_section)]
const _: &'static str = include_str!("./types.d.ts");
```
```rust
const TS_INTERFACE_EXPORT: &str = "interface Person2 { height: number; }";
#[wasm_bindgen(typescript_custom_section)]
const _: &str = TS_INTERFACE_EXPORT;
```

this pr is inspired by [#3308](https://github.com/rustwasm/wasm-bindgen/pull/3308) , and will fix [#2828]

this pr is compatible with current encode logic, so `wasm-bindgen cli` can decode normally without updating

### CODE REVIEW

**Comment by swnb** [comment]
I don't mean to rush you , but any progress about this pr @daxpedda

**Comment by daxpedda** [comment]
Unfortunately I'm not really qualified to review any TS related stuff, so I leave it to other maintainers.

**Comment by swnb** [comment]
all right, but I think you are the only maintainer who responded positively.

File: crates/typescript-tests/src/custom_section_types
Comment (swnb) [reply crates/typescript-tests/src/custom_section_types:1]: Ok

File: src/lib.rs
Comment (swnb) [reply src/lib.rs:1838]: sorry about that , i was meant to use `mem::MaybeUninit`, but i write wrong method, anyway, i will change it and avoid unsafe

File: crates/cli-support/src/decode.rs
Comment (swnb) [reply crates/cli-support/src/decode.rs:15]: i will change it to `LitOrExpr`

File: crates/shared/src/lib.rs
Comment (swnb) [reply crates/shared/src/lib.rs]: I will add detailed documentation

File: src/lib.rs
Comment (swnb) [reply src/lib.rs:1857]: i will add some comment;

Comment (swnb) [reply src/lib.rs]: I've commented the code, but is there a need to calculate the length of var-length-integer during compilation? I personally think it is not necessary, so I directly use the maximum length here. If you think it is necessary, I can add this logic. I think the impact on the size of the compiled result will be minimal.

File: src/lib.rs
Comment (Liamolucko) [inline src/lib.rs:1824]: I don't think these nested submodules are particularly necessary, putting these functions directly into `__rt` is fine.

File: src/lib.rs
Comment (Liamolucko) [reply src/lib.rs]: Yeah, I agree that it isn't really necessary, it shouldn't make much difference either way.

File: src/lib.rs
Comment (swnb) [reply src/lib.rs:1824]: Ok

File: src/lib.rs
Comment (Liamolucko) [inline src/lib.rs:1858]: This isn't really a TODO if we've decided not to do it.

**Review by Liamolucko — APPROVED** [review]
Thanks!

Could you add a changelog entry?

File: src/lib.rs
Comment (swnb) [reply src/lib.rs:1858]: Ok，i will remove TODO;

File: CHANGELOG.md
Comment (Liamolucko) [inline CHANGELOG.md:8]:
```suggestion
* Added support for arbitrary expressions when using `#[wasm_bindgen(typescript_custom_section)]`.
```

**Pull request merged**
